akka-actor/src/main/scala/akka/util/WildcardTree.scala

```diff
@@ -18,15 +18,26 @@ private[akka] final case class WildcardTree[T](data: Option[T] = None, children:
       copy(data = Some(d))
     } else {
       val e = elems.next()
-      copy(children = children.updated(e, children.get(e).getOrElse(WildcardTree()).insert(elems, d)))
+      copy(children = children.updated(e, children.getOrElse(e, WildcardTree()).insert(elems, d)))
     }
 
-  @tailrec final def find(elems: Iterator[String]): WildcardTree[T] =
-    if (!elems.hasNext) this
-    else {
-      (children.get(elems.next()) orElse children.get("*")) match {
-        case Some(branch) ⇒ branch.find(elems)
-        case None         ⇒ WildcardTree()
+  def find(elems: Iterator[String]): WildcardTree[T] = {
+    case class BFSState(wildcardTree: WildcardTree[T], elems: List[String])
+    val queue = mutable.Queue(BFSState(this, elems.toList))
+    while (queue.nonEmpty) {
+      val state = queue.dequeue()
+      if (state.elems.isEmpty) return state.wildcardTree
+      else {
+        val kids = state.wildcardTree.children
+
+        kids get state.elems.head foreach explore
+        kids get "*" foreach explore
+        kids get "**" foreach { kid ⇒ explore(kid); exploreYourself() }
+
+        def explore(kid: WildcardTree[T]) = queue.enqueue(BFSState(kid, state.elems.tail))
+        def exploreYourself() = queue.enqueue(state.copy(elems = state.elems.tail))
       }
     }
+    WildcardTree()
+  }
```

Comment:
I am fine if the normal case is not much affected and the generic case is somewhat slower. It would be nice to see how much slower actually.